exchange/tx.go

```diff
@@ -269,26 +269,45 @@ func (tx *Tx) buildRoot() error {
 
 // Ref returns the DataRef associated with this transaction
 func (tx *Tx) Ref() *DataRef {
-	return &DataRef{
+	ref := &DataRef{
 		PayloadCID:  tx.root,
 		StoreID:     tx.storeID,
 		PayloadSize: tx.size,
 	}
+
+	for key := range tx.entries {
+		ref.Keys = append(ref.Keys, []byte(key))
+	}
```

Comment:
I think this is find, but we have to make sure this method isn't called when the entries haven't been loaded into memory

Reply:
Yeah you could check the length of `tx.entries` first and if it's empty, run `MapKeys`